ruby/ql/lib/codeql/ruby/frameworks/core/internal/IOOrFile.qll

```diff
@@ -0,0 +1,242 @@
+  }
+
+  /** Gets the API used to perform this call, either "IO" or "File" */
+  abstract string getApi();
+
+  /** DEPRECATED: Alias for getApi */
+  deprecated string getAPI() { result = this.getApi() }
+
+  /** Gets a node representing the data read or written by this call */
+  abstract DataFlow::Node getADataNodeImpl();
+
+  /** Gets a string representation of the receiver kind, either "class" or "instance". */
+  abstract string getReceiverKind();
+}
+
+// Match some simple cases where a path argument specifies a shell command to
+// be executed. For example, the `"|date"` argument in `IO.read("|date")`, which
+// will execute a shell command and read its output rather than reading from the
+// filesystem.
+predicate pathArgSpawnsSubprocess(Expr arg) {
+  arg.getConstantValue().getStringlikeValue().charAt(0) = "|"
+}
+
+/**
+ * A method call that performs a read using either the `IO` or `File` classes.
+ */
+class IOOrFileReadMethodCall extends IOOrFileMethodCall {
+  private string api;
+  private string receiverKind;
+
+  IOOrFileReadMethodCall() {
+    exists(string methodName | methodName = this.getMethodName() |
+      // e.g. `{IO,File}.readlines("foo.txt")`
+      receiverKind = "class" and
+      methodName = ["binread", "foreach", "read", "readlines"] and
+      api = ["IO", "File"] and
+      this = API::getTopLevelMember(api).getAMethodCall(methodName)
+      or
+      // e.g. `{IO,File}.new("foo.txt", "r").getc`
+      receiverKind = "interface" and
+      (
+        methodName =
+          [
+            "getbyte", "getc", "gets", "pread", "read", "read_nonblock", "readbyte", "readchar",
+            "readline", "readlines", "readpartial", "sysread"
+          ] and
+        (if this.getReceiver() instanceof File::FileInstance then api = "File" else api = "IO")
+      )
+    )
+  }
+
+  override string getApi() { result = api }
+
+  /** DEPRECATED: Alias for getApi */
+  deprecated override string getAPI() { result = this.getApi() }
+
+  override DataFlow::Node getADataNodeImpl() { result = this }
+
+  override string getReceiverKind() { result = receiverKind }
+}
+
+/**
+ * A method call that performs a write using either the `IO` or `File` classes.
+ */
+class IOOrFileWriteMethodCall extends IOOrFileMethodCall {
+  private string api;
+  private string receiverKind;
+  private DataFlow::Node dataNode;
+
+  IOOrFileWriteMethodCall() {
+    exists(string methodName | methodName = this.getMethodName() |
+      // e.g. `{IO,File}.write("foo.txt", "hello\n")`
+      receiverKind = "class" and
+      api = ["IO", "File"] and
+      this = API::getTopLevelMember(api).getAMethodCall(methodName) and
+      methodName = ["binwrite", "write"] and
+      dataNode = this.getArgument(1)
+      or
+      // e.g. `{IO,File}.new("foo.txt", "a+).puts("hello")`
+      receiverKind = "interface" and
+      (if this.getReceiver() instanceof File::FileInstance then api = "File" else api = "IO") and
```

Comment:
I think that the test failure is due to the change from
```
      (
        this.getReceiver() = ioInstance() and api = "IO"
        or
        this.getReceiver() = fileInstance() and api = "File"
      ) and
``` 
to
```
      (if this.getReceiver() instanceof File::FileInstance then api = "File" else api = "IO") and
```

The receiver of the call to `<<` at `logging.rb:26:18:26:34` is a `Logger` instance, but this change will make this be interpreted as a call against an `IO` instance due to not restricting the receiver type in the `else` branch.

This test case is intended to be picked up as a cleartext logging instance rather than as a cleartext storage instance.

Reply:
Thanks! You're right, that was the problem.